### PR TITLE
Prevent coalescing for web source/trigger headers

### PR DESCRIPTION
**Pull request by agarant**
fixes: https://github.com/WICG/attribution-reporting-api/issues/1202

- Prevent header coalescing when receiving the web registration headers.
- Reject the headers if more than 1 is received.

***
<a href="https://pr-preview.s3.amazonaws.com/agarant/attribution-reporting-api/pull/1212.html" title="Last updated on Mar 27, 2024, 3:39 PM UTC (30f5805)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/WICG/attribution-reporting-api/1212/7e8b4c9...agarant:30f5805.html" title="Last updated on Mar 27, 2024, 3:39 PM UTC (30f5805)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -1838,6 +1839,18 @@ a [=header value=] or null |webTriggerHeader|, a [=header value=] or null
...
+and a [=header name=] header:
+1. Let |non-coalesced-headers| be the result of [=get non-coalesced headers=] with |headers| and |header|.
+1. If |non-coalesced-headers| [=list/is empty=], return null.
+1. Return |non-coalesced-headers|[0].
+
+To <dfn export>get non-coalesced headers</dfn> given a [=response/header list=] |headers|,
+and a [=header name=] header:
+1. If |headers| does not [=header list/contain=] |header|, return a new [=list=].
+1. Return the [=header value|values=] of all [=header|headers=] in |headers| whose [=header name|name=]
+    is a [=byte case insensitive=] match for |header|, in order.
```
Comment (apasel422) [inline index.bs:1852]:
```suggestion
    is a [=byte-case-insensitive=] match for |header|, in order.
```

File: index.bs [hunk 2]
```diff
@@ -1838,6 +1839,18 @@ a [=header value=] or null |webTriggerHeader|, a [=header value=] or null
 
      </dl>
 
+To <dfn export>get the first non-coalesced header</dfn> given a [=response/header list=] |headers|,
+and a [=header name=] header:
+1. Let |non-coalesced-headers| be the result of [=get non-coalesced headers=] with |headers| and |header|.
+1. If |non-coalesced-headers| [=list/is empty=], return null.
+1. Return |non-coalesced-headers|[0].
+
+To <dfn export>get non-coalesced headers</dfn> given a [=response/header list=] |headers|,
```
Comment (apasel422) [inline index.bs:1848]: I don't think we want to export these algorithms; ideally they'd ultimately be part of the `fetch` spec itself.

File: index.bs [hunk 3]
```diff
@@ -1838,6 +1839,18 @@ a [=header value=] or null |webTriggerHeader|, a [=header value=] or null
 
      </dl>
 
+To <dfn export>get the first non-coalesced header</dfn> given a [=response/header list=] |headers|,
```
Comment (apasel422) [inline index.bs:1842]:
```suggestion
To <dfn export>get the first non-coalesced header</dfn> given a [=response/header list=] |headers|
```

File: index.bs (hunk 2 above)
Comment (apasel422) [inline index.bs:1848]:
```suggestion
To <dfn export>get non-coalesced headers</dfn> given a [=response/header list=] |headers|
```

File: index.bs [hunk 4]
```diff
@@ -1838,6 +1839,18 @@ a [=header value=] or null |webTriggerHeader|, a [=header value=] or null
...
+and a [=header name=] header:
+1. Let |non-coalesced-headers| be the result of [=get non-coalesced headers=] with |headers| and |header|.
+1. If |non-coalesced-headers| [=list/is empty=], return null.
+1. Return |non-coalesced-headers|[0].
+
+To <dfn export>get non-coalesced headers</dfn> given a [=response/header list=] |headers|,
+and a [=header name=] header:
+1. If |headers| does not [=header list/contain=] |header|, return a new [=list=].
```
Comment (apasel422) [inline index.bs:1850]:
```suggestion
```
There's no reason to explicitly handle this case, since the subsequent step will return an empty list.

**Review by apasel422 — CHANGES_REQUESTED** [review]
Can we rename the PR to something like `Prevent coalescing for web source/trigger headers`?

File: index.bs (hunk 1 above)
Comment (agarant) [reply index.bs:1852]: It is defined at the top with "byte case insensitive".

Comment (apasel422) [reply index.bs:1852]: Right, but that's not how it's spelled in the fetch spec: https://infra.spec.whatwg.org/#byte-case-insensitive

Comment (agarant) [reply index.bs]: I removed the mapping and use the link directly instead.

File: index.bs [hunk 5]
```diff
@@ -1840,10 +1840,10 @@ a [=header value=] or null |webTriggerHeader|, a [=header value=] or null
 
      </dl>
 
-To <dfn>get the first non-coalesced header</dfn> given a [=response/header list=] |headers|
+To <dfn>get non-coalesced header</dfn> given a [=response/header list=] |headers|
```
Comment (apasel422) [inline index.bs:1843]: `get non-coalesced header if unique` might be more clear.

File: index.bs [hunk 6]
```diff
@@ -1860,10 +1871,10 @@ To <dfn export>process an attribution eligible response</dfn> given a [=suitable
     "<code>[=eligibility/event-source-or-trigger=]</code>".
 1. Let |reportingOrigin| be |response|'s [=response/URL=]'s [=url/origin=].
 1. If |reportingOrigin| is not [=check if an origin is suitable|suitable=], return.
-1. Let |sourceHeader| be the result of [=header list/get|getting=]
+1. Let |sourceHeader| be the result of [=get non-coalesced header|getting=]
     "`Attribution-Reporting-Register-Source`" from |response|'s
     [=response/header list=].
-1. Let |triggerHeader| be the result of [=header list/get|getting=]
+1. Let |triggerHeader| be the result of [=get non-coalesced header|getting=]
```
Comment (apasel422) [inline index.bs:1877]: I don't think this properly handles:

```http
Attribution-Reporting-Register-Source: {...}
Attribution-Reporting-Register-Source: {...}
Attribution-Reporting-Register-Trigger: {...}
```
In this case, the two sources will be ignored but the trigger will be handled. IMO this should be an error that causes all the headers to be ignored instead.

**Review by apasel422 — CHANGES_REQUESTED** [review]

**Review by apasel422 — APPROVED** [review]

**Pull request merged**